### DISCUSSION TITLE
Major data loss using jspsych / pavlovia

### DISCUSSION THREAD

**Discussion by Cluster-Bomb** [jspsych/jsPsych#2032] [Q&A]
URL of experiment: [Effect of Frame of Reference on Biological Motion ](https://run.pavlovia.org/Sean_Mulligan/the-effect-of-priming-on-frame-of-reference-while-viewing-biological-motion-)

Hey there, 
This is a question for anyone who uses Pavlovia. Am trying to get help on the Pavlovia discussion forums but thought I would try here as well. 

Description of the problem: About 50% of participant data is not saving - despite “save incomplete results” (Pavlovia option of saving partial data from participants) being on. Can’t replicate the issue with despite trying with five other people using various browser combinations. Both complete experiments and incomplete experiments were saved during testing. The participants are are funneled from SONA for credits and they are completing the experiment (confirmed through contact) but no data file is created. My data files that do save are 99kb so it is unlikely an internet connection problem. I notice that this is a recurring problem that many Pavlovia/ jspsych users have to varying degrees:

1. [Two experiments 15-20% data loss](https://discourse.psychopy.org/t/pavlovia-and-jspsych-data-loss/14443)
2. [Three other experiments - one with 40% data loss](https://discourse.psychopy.org/t/issues-with-data-missing-or-lost/14433)
3. [10% of data not saved in several experiments in lab](https://ccbrain.org/science/Onlinetesting/)
4. [Two experiments data not saved at all](https://discourse.psychopy.org/t/data-not-saving-to-pavlovias-servers-jspsych/13621)
5. [Multi-part experiments losing data](https://discourse.psychopy.org/t/jspsych-saving-options-on-pavlovia/12880)

Some discussion has been around creating a delay in order to ensure the results are saved to Pavlovia before jspsych completes - but thomas pronk in the first discussion link above suggests that this is not required. In addition the partial save option in Pavlovia should have captured the data anyway. 

Any help would be appreciated
     
                   Cheers
                                Sean

Here are the relevent code fragments: 

**Plugin version**
<script type="text/javascript" src="https://lib.pavlovia.org/jspsych-pavlovia-3.2.0.js"></script>`

**Pavlovia init** 

```html
<body>
	<script type='text/javascript'>

		/* create timeline */
		var timeline = [];


		/* init connection with pavlovia.org */
		var pavlovia_init = {
			type: "pavlovia",
			command: "init"
		};
		timeline.push(pavlovia_init);
		
		/* generate a random subject ID that contains 8 alphanumeric characters*/
        var rand_subject_id = jsPsych.randomization.randomID(8);
        
        /* add the ID to the data for all trials */
        jsPsych.data.addProperties({ID: rand_subject_id});
        
        /* get SONA is from URL*/
        let sona_id = jsPsych.data.urlVariables()['sona_id']
        
        /* add the SONA ID to the data for all trials */
        jsPsych.data.addProperties({SONA_ID: sona_id});
```
**End of experiment**

```html
/ exit fullscreen mode
    timeline.push({
      type: 'fullscreen',
      fullscreen_mode: false
    });
    
	var debrief_block = {
      type: "html-keyboard-response",
      stimulus: function() {

        var trials = jsPsych.data.get().filter({task: 'response'});
        var correct_trials = trials.filter({correct: true});
        var accuracy = Math.round(correct_trials.count() / trials.count() * 100);
        var rt = Math.round(correct_trials.select('rt').mean());

        return `<p>You responded correctly on ${accuracy}% of the trials.</p>
          <p>Your average response time was ${rt}ms.</p>
          <p>Press any key to complete the experiment. Thank you!</p>`;
        }
    };
		timeline.push(debrief_block);

		/* finish connection with pavlovia.org */
		var pavlovia_finish = {
			type: "pavlovia",
			command: "finish"
		};
		timeline.push(pavlovia_finish);


		/* start the experiment */
		jsPsych.init({
			timeline: timeline,

			on_finish: 
			
			function (data) {
                window.location.assign("https://uws.sona-systems.com/webstudy_credit.aspx?experiment_id=1513&credit_token=91266d13e0d142c0b046b13ec4c88b91&survey_code="+sona_id)
            },
            
		});
```

**Comment by ghost** [comment]
Hello @Cluster-Bomb ,

I hope that you are having a great week so far and a Happy New Year!

I was wondering if you managed to find a solution to the data loss? I am having a similar problem (more than 50% of my participants' data are lost), but I am integrating my jsPysch experiment with SONA and saving the data via Firebase. 

It's strange because when I am testing the experiment locally, the data saves completely with Firebase. However, once I integrate SONA this results in a massive data loss...

**Comment by Cluster-Bomb** [comment]
Hi there,

Never did find out why the data loss was so big.
Using the latest jsPaych and plugin versions maybe reduced it 5-10%.
Putting a redirect back to SONA maybe accounted for another 5%. In the end
I had jspsych save a local copy of the results and got the participants to
email it to me. While only half the participants actually sent it back -
this did reduce the data loss to acceptable levels.

Cheers
            Sean

**Reply by ghost** [reply]
Hi Sean,

Thank you very much for your reply, and I am sorry that the data loss problems still persisted. However, I'm happy to hear that this was somewhat resolved by having participants send the local copy of their results.

I am not sure how helpful this is since I am not using Pavlovia, but I was able to reduce the data loss that resulted from the participants being redirected to SONA by using a JavaScript promise. Here is the [link](https://github.com/jspsych/jsPsych/discussions/2435) to the discussion thread, in case this is useful (I have included my solution at the bottom).

Happy Saturday!

Sincerely,
Michelle
